Command line arguments for specifying options when adding torrents

briankendall opened the pull request:
This pull request adds the following command line arguments to qBittorrent that apply when specifying new torrents via the command line:

-p | --path <path>: Torrent save path
--add-started | --add-paused: Add torrents as started or paused
--skip-hash-check: Skip hash check
--category <category name>: Assign torrents to category
--sequential: Download torrents in sequential order

Using one or more of these arguments will make qBittorrent skip the "Add new torrent" dialog and simply add the torrent with the specified options, falling back on the application-wide settings for the options not specified. This fulfills many users' requests for a better method of adding torrents via scripts or some other method of automation, including the ability to specify options for those new torrents, and in particular being able to specify what path they will be downloaded to.

glassez commented on src/app/main.cpp:
```diff
@@ -302,6 +352,38 @@ QBtCommandLineParameters parseCommandLine()
+            else if (arg == QLatin1String("--add-started")) {
+                result.addPaused = false;
+            }
+            else if (arg == QLatin1String("--add-paused")) {
```
Having both add-started and add-paused params is bad idea. One of them is redundant.

glassez commented on src/app/main.cpp:
```diff
@@ -302,6 +352,38 @@ QBtCommandLineParameters parseCommandLine()
             }
 #endif
+            else if ((arg == QLatin1String("-p")) || (arg == QLatin1String("--path"))) {
+                if ((i+1) < appArguments.size()) {
```
Here and in similar places you break coding style rule. You must insert spaces around arithmetic operators.

briankendall commented:
> Having both add-started and add-paused params is bad idea. One of them is redundant.

I've replaced it with a single argument, --status=[started|paused]

Since this pull request adds some new arguments that specify a value in addition to the existing one (--webui-port), I factored out the logic for parsing them and made it a bit more flexible. These arguments can now be followed by an equal sign (=) or a space. This fixes an inconsistency in my earlier commit where --webui-port required an equal sign but the ones I added required a space.

> Here and in similar places you break coding style rule. You must insert spaces around arithmetic operators.

Damn! I was really trying to follow all of the code styling rules, but I slipped up in that one place. Hopefully I got it right this time.

glassez commented:
@briankendall you should squash all these commits into one (there are not so many changes to have multiple commits in the qBittorrent history). You can leave it as is now but the maintainer will ask you doing it before merging.

briankendall commented:
> you should squash all these commits into one

Done.

briankendall commented:
I just cleaned up the code style a bit more (was using curly braces for one-line if statements) and added one more command line argument for downloading first and last pieces first.

wiiaboo commented on src/app/application.cpp:
```diff
@@ -379,15 +379,58 @@ void Application::processParams(const QStringList &params)
 
     foreach (QString param, params) {
         param = param.trimmed();
+        
```
You should also remove the trailing spaces in this and other lines.

briankendall replied:
It is done.

wiiaboo commented:
I've been using qBittorrent master with this patch for a few weeks with no issues.
I can confirm it also works with 3.3.7. Both on Windows 10 64-bit.
This is a very useful feature for use with scripts and other programs.

briankendall commented:
Any chance of this being merged in soon? I've been maintaining my own builds of qBittorrent for myself and a few friends that has this feature compiled in, and would love it if I didn't have to do that anymore. It's been working great for me, my aforementioned friends, and @wiiaboo.

MatrixView commented:
Please, please, please... merge this. It's been over 6 months and this solves many related requests. The patch is pretty much ESSENTIAL for automation/scripting users...

uncleungie commented:
Please merge this! It's an extremely useful feature for scripting and automation. I can't use qBittorrent without it.

sledgehammer999 commented:
@briankendall I think this will conflict (source-wise) with #5214 which I currently am reviewing. I am waiting for the author to rebase.
Can you kindly watch it, and when I merge it, rebase this if needed and remind me to review it too?

wiiaboo commented:
@briankendall nice to finally get some recognition this PR exists. Could you merge this in the same release as #5214, at least? Getting kinda tired of rebuilding qbt just to have this feature. I'd just give up on keeping up-to-date if merging is non-trivial.

briankendall commented:
@sledgehammer999 Quite right, that will surely cause conflicts.

> Can you kindly watch it, and when I merge it, rebase this if needed and remind me to review it too?

Will do.

briankendall commented:
@sledgehammer999 Now that #5214 is merged, I've rebased using latest master. (In doing so I cleaned up the formatting for the command line usage text again -- I just can't abide poorly formatted text! 😜) Let me know if things look good or if there's further changes to be made. I'm hoping we can finally get this merged soon!

LordNyriox commented:
@sledgehammer999 approveFeature

LordNyriox commented:
@glassez, @evsh, @Chocobo1:  Review?

briankendall commented:
Resolved latest merge conflicts.

glassez commented:
@evsh, you recently had to deal with handling command line parameters. I think you have something to say here.

evsh commented on src/app/options.cpp:
```diff
@@ -247,6 +252,12 @@ namespace
     constexpr const StringOption CONFIGURATION_OPTION = {"configuration"};
     constexpr const BoolOption PORTABLE_OPTION = {"portable"};
     constexpr const BoolOption RELATIVE_FASTRESUME = {"relative-fastresume"};
+    constexpr const StringOption SAVE_PATH_OPTION = {"path"};
```
I would avoid such a general name for the parameter. Who knows what other path we would add in the future?

LordNyriox replied:
@briankendall, @evsh:  How about `torrent-path`?

glassez replied:
We use "save path" term in this meaning, so it should be "save-path" here.

briankendall replied:
Good point. I'll change it to save-path unless you have a different suggestion.

evsh commented on src/app/options.cpp:
```diff
@@ -247,6 +252,12 @@ namespace
     constexpr const BoolOption PORTABLE_OPTION = {"portable"};
     constexpr const BoolOption RELATIVE_FASTRESUME = {"relative-fastresume"};
+    constexpr const StringOption SAVE_PATH_OPTION = {"path"};
+    constexpr const StringOption STATUS_OPTION = {"status"};
```
Same here.

LordNyriox replied:
@briankendall, @evsh:  Likewise, how about `torrent-status`?

glassez replied:
We use "add paused" term in this meaning, so it should be "add-paused" here.

briankendall replied:
Similarly will change it to torrent-status

glassez replied:
IMO, "status" is incorrect in this context.

evsh replied:
I would use bool option `--add-paused`, which will be shorter without the value part than `--status=paused`. Are there any more statuses except these two?

briankendall replied:
@evsh `--add-paused` doesn't give the option of someone adding a torrent unpaused when the app is configured to add all torrents paused by default, so it ought to include both options.

evsh replied:
Maybe we better extend the `BoolOption` class to accept a value? To make it possible to pass `--add-paused=false`.

briankendall replied:
I spent some time thinking about what the name of this option should be, and settled on `status` (now `torrent-status`) because I can't find any other terminology used elsewhere in the app that refers to the attribute of a torrent that distinguishes whether it's paused or not in the way that we need.

In the add torrent dialog, there is a check box that reads "Start torrent", but since we're overriding the default behavior with this command line option, a simple toggle isn't good enough. The user needs to specify one of three states: use default settings, pause torrent, start torrent. The `torrent-status` approach allows for all three.

The actual setting in the preferences for specifying the default behavior is "Do not start the download automatically" which doesn't really help for coming up with an intuitive command line option either.

In the main window, whether a torrent is paused or running is indicated in the status column.

So I eventually settled with `torrent-status`. I originally set it up so that there was an `--add-paused` and `--add-started` option but a while back @glassez suggested that was redundant and I'm inclined to agree. I think `torrent-status=paused|started` is the way to go.

glassez replied:
>I think `torrent-status=paused|started` is the way to go.

But what not `--add-paused=true|false`?

briankendall replied:
@glassez Yes, that could work. I originally discounted that as BoolOption doesn't current allow a way to specify a value, and I don't think it should since it's implementing a specific common kind of command line option. Instead I could continue using StringOption and checking its value, requiring it to be true or false. That would be the easiest and fastest change.

Alternatively, I could add a new Option-derived class called something like TriStateBoolOption, that requires explicitly specifying a value. That would allow some flexibility in how the value is specified, i.e. true, false, 1, 0, on, and off could all be valid values.

evsh replied:
I would model `BoolOption` after the same entity in `boost::program_options`, i.e. give it ability to parse supplied value and set an implicit one, which is taken when no explicit value is provided.
If  you, @briankendall, do not want to do that or wait for such implementation, IMO `StringOption` with a custom check would be fine. Let's just avoid option name change in the future.

briankendall replied:
@evsh I don't mind implementing this, as I do prefer doing things "the right way". That said I don't think doing something like `boost::program_options` like you describe will work here, as I need an unspecified option to have a value that explicitly means it was undefined rather than an implicit true / false value. So think it would be better to utilize the existing `TriStateBool` class (which exists for exactly that purpose) and create a new kind of option based on that.

evsh replied:
@briankendall, excuse me, don't understand yet why do you need the third value. Why just absent parameter in the command line does not work? Sure, you need the tri-state value further on, bu why it is needed in the command line?

briankendall replied:
@evsh It's mainly because, without having an `Option`-derived class that can return a tri-state bool, the mechanism where options can be specified using environment variables no longer works.

When initializing a value in the constructor of `QBtCommandLineParameters` you need to use something like: `addPaused(PAUSED_OPTION.value(env))`

This will initialize the variable `addPaused` to either true or false. So if we don't get a value from the corresponding environment variable, `addPaused` is initialized to false rather than to undefined. And now we don't know whether the user actually specified false in the environment variable or didn't specify anything at all. Using `TriStateBoolOption` fixes this.

On top of that, and perhaps even more importantly, I think having a separate class called `TriStateBoolOption` is *conceptually* correct as it provides a different function (namely the ability to have an undefined state) in comparison to `BoolOption`. Better to make that explicit in the definition of an option than to have its behavior not come from its definition but its implementation details. You don't want people to have to check the code carefully to know that `add-paused` behaves differently than every other `BoolOption`.

glassez replied:
I agree with @briankendall we need some way for "use application default value for addPaused". If user omit corresponding parameter this shouldn't mean "false".

evsh commented on src/app/options.cpp:
```diff
@@ -312,6 +366,27 @@ QBtCommandLineParameters parseCommandLine(const QStringList &args)
+            else if (arg == STATUS_OPTION) {
+                result.torrentStatus = STATUS_OPTION.value(arg);
+                if (result.torrentStatus != "paused" && result.torrentStatus != "started")
+                    throw CommandLineParameterError(QObject::tr("%1 must specify either started or paused.")
```
Either remove "started" and "paused" from the translatable string or make it clear for a translator that they have to be kept as is.

briankendall replied:
Good catch. I think `started` and `paused` had better remain consistent regardless of system language otherwise scripts that execute in one language will fail in another.

evsh commented on src/app/options.cpp:
```diff
@@ -343,43 +418,86 @@ const QString& CommandLineParameterError::messageForUser() const
+    return lines.join("\n");
+}
+
+QString wrapOptionText(const QString &text)
```
Don't quite understand why we can pass `WRAP_AT_COLUMN` as default value, but can not do the same with `USAGE_TEXT_COLUMN`?

briankendall replied:
I'll simplify things so that there's just one function, `wrapText`.

briankendall commented:
Just updated my branch. Changes include:

- `status=<started|paused>` is now `add-paused=<true|false>`
- `path` is now `save-path`
- Combined `wrapOptionText` and `wrapText` into one function
- Fixed inconsistent use of translatable strings
- Added directions for translators when using English words is necessary in the translation
- Added `TriStateBoolOption` class for implementing `add-paused`
- Added `skip-dialog` option as per @evsh 's suggestion

evsh commented on src/app/application.cpp:
```diff
@@ -403,15 +403,63 @@ void Application::processParams(const QStringList &params)
+        // any of these should prevent the "Add new torrent" dialog from
+        // opening:
+
+        if (param.startsWith("@path=")) {
```
It is still "path" here.

evsh commented on src/app/application.cpp:
```diff
@@ -403,15 +403,63 @@ void Application::processParams(const QStringList &params)
+
+        if (param.startsWith("@path=")) {
+            torrentParams.savePath = param.mid(6);
+            skipTorrentDialog = true;
```
Why this is still here?

briankendall replied:
Because utilizing any of the command line options relating to adding a torrent prevents the "Add new torrent" dialog from opening.

evsh commented on src/app/options.cpp:
```diff
@@ -235,6 +240,69 @@ namespace
+        
+        QString usage() const
+        {
+            return StringOption::usage("true|false");
```
Please use `QLatin1String`

evsh commented on src/app/options.cpp:
```diff
@@ -235,6 +240,69 @@ namespace
+                return TriStateBool::False;
+            }
+            else {
+                //: Translation must use English "true" and "false"
```
Will this go into Transifex?

briankendall replied:
I haven't used Transifex before so I'm not sure. But this is the officially supported way of adding comments to translatable strings according to Qt's documentation. Is there another method I should use?

evsh commented on src/app/options.cpp:
```diff
@@ -235,6 +240,69 @@ namespace
+            if (val.isEmpty()) {
+                return TriStateBool::Undefined;
+            }
+            else if (val.toUpper() == "TRUE" || val == "1") {
```
`QLatin1String`

evsh commented on src/app/options.cpp:
```diff
@@ -235,6 +240,69 @@ namespace
+                return TriStateBool::False;
+            }
+            else {
+                //: Translation must use English "true" and "false"
```
Will this go into Transifex?

evsh commented on src/app/options.cpp:
```diff
@@ -247,6 +315,13 @@ namespace
+    constexpr const BoolOption SKIP_HASH_CHECK_OPTION = {"skip-hash-check"};
+    constexpr const StringOption CATEGORY_OPTION = {"category"};
+    constexpr const BoolOption SEQUENTIAL_OPTION = {"sequential"};
+    constexpr const BoolOption FIRST_AND_LAST_OPTION = {"first-and-last"};
```
I don't understand why `BoolOption` is OK here, but `TriStateBoolOption` is needed for 'add-paused'? Don't they behave identically in terms of absent values? I.e. when value in command line is absent, program setting has to be used.

briankendall replied:
There is no setting for "download first and last piece first" in the application-wide preferences. It always defaults to off so therefore there's no need for a tri-state bool.

evsh replied:
OK.

glassez replied:
This param won't affect if we add magnet, isn't it?

briankendall replied:
@glassez You're right, it doesn't have an effect in that case. I'm not sure why, but the issue seems to be somewhere within `BitTorrent::Session::addTorrent()` and/or everything it calls into. I don't think I'm the right person to debug this one.

glassez replied:
It's not a bug. We can't prioritize something since we haven't any torrent metadata in case we add it using magnet link. We can do it deferred when metadata has to be loaded. Please leave TODO comment near corresponding AddTorrentParams member.

briankendall replied:
OK. Adding a magnet link / torrent URL does work with the "download in sequential order" feature. I'll see if I can't figure out how to get it to work with "first and last piece priority" and failing that I'll add the TODO.

briankendall replied:
@sledgehammer999 What about cases where qBittorrent's application-wide settings are configured to a) add torrents paused by default or b) skip the dialog by default? There should be a way to override the application settings from the command line and force adding a torrent in a started state or force showing the "Add new torrent" dialog, hence the ability to specify `--add-paused=false` or `--skip-dialog=false`.

We eventually settled on having a tri-state bool command line option since it seemed like the best solution, especially since the options for whether a new torrent should be started or paused are mutually exclusive.

sledgehammer999 replied:
Maybe I have misunderstood the code or my comment was incomplete.
Why isn't it possible for user to use `--add-paused`, `--add-paused=true`, `--add-paused=false`?
In this case the omission of a value should be equal to `true`. And if the user doesn't use the command switch the application wide setting should apply.

briankendall replied:
Oh, I see what you're getting at. You're correct, we could make it legal to omit a value, and have it default to what the option's name implies.

I could implement that by adding an additional argument to the constructor of `TriStateBoolOption` that specifies what it defaults to without a value and then change it to not  throw an error and use that value when no argument is specified.

briankendall replied:
That said, I don't think it's the end of the world to require a value, and I am keen to see this PR completed! Changing `TriStateBoolOption` option to work that way isn't a trivial amount of work.

sledgehammer999 replied:
Okay, I'll merge it then. I'll try to make the change myself today. If you see nothing by tomorrow, feel free to open a followup PR.

evsh commented on src/app/options.cpp:
```diff
@@ -264,7 +339,54 @@ QBtCommandLineParameters::QBtCommandLineParameters(const QProcessEnvironment &en
+    // torrent paths or URLs.
+
+    if (!savePath.isEmpty())
+        result.append(QString("@path=%1").arg(savePath));
```
"path"

evsh commented on src/app/options.cpp:
```diff
@@ -312,6 +434,27 @@ QBtCommandLineParameters parseCommandLine(const QStringList &args)
+                result.savePath = SAVE_PATH_OPTION.value(arg);
+            }
+            else if (arg == PAUSED_OPTION) {
+                result.addPaused = PAUSED_OPTION.value(arg);
```
So, continuing, my idea is: only `result.addPaused` has to be of type `TriStateBool`. The option itself is a bool option which accepts value and can be omitted. The third case is here already, in this very test.

briankendall replied:
You're forgetting about environment variables being used to specify options. It's definitely necessary to have `TriStateBoolOption` in order for that to work. Not to mention there's my earlier comment about how using `TriStateBoolOption` unambiguously informs people reading the code that it behaves differently than the other bool options, which I think vastly improves comprehensibility.

evsh replied:
You are right, option object with optional value (sorry) is certainly a cleaner solution then.

evsh commented on src/app/application.cpp:
```diff
@@ -403,15 +403,63 @@ void Application::processParams(const QStringList &params)
+
+        if (param.startsWith("@savePath=")) {
+            torrentParams.savePath = param.mid(10);
+            skipTorrentDialog = true;
```
Why this assignment is still present?

briankendall replied:
Because specifying a save path prevents the new torrent dialog from opening.

evsh replied:
Can't agree with you. IMO, it would be perfectly fine to supply a save path, a magnet uri and expect to see the dialog. And that is why I asked you to add the option to control it. It is perfectly fine to skip the dialog by default, but where is `skipTorrentDialog = false` then?

briankendall replied:
Unfortunately the save dialog currently has no facility for passing in torrent options. Meaning if I use command line arguments such as `--save-path`, `--category`, `--start-paused` and so on, those options are not reflected in the save dialog and end up having no effect. They only work when adding a torrent using `BitTorrent::Session::instance()->addTorrent()`.

The thing is that these command line arguments are mainly intended to be used for automating qBittorrent. In those instances it's not desirable to have the save dialog open, hence why I set it up so that specifying any of them prevents it from opening.

I thought adding a `--skip-dialog` option was useful, though, because someone might want to add a torrent, skip the dialog, and use the application settings without overriding anything, which wasn't possible previously.

In order to accomplish what you're talking about, I'd have to make significant changes to `AddNewTorrentDialog` so that it can accept an optional `BitTorrent::AddTorrentParams` parameter and properly pass those settings on when adding the torrent, even the ones that can't be specified by the dialog itself. I'm not very familiar with `AddNewTorrentDialog` so it would take some work and have the potential for introducing new bugs. I'd prefer not to do that unless there's a compelling reason for it.

evsh replied:
Ah! What a pity! Could you leave a comment or TODO marker somewhere then regarding this problem?

briankendall replied:
Sure.

evsh replied:
Thank you!

briankendall replied:
Actually this may require some future proofing. If someone were to make the change we're discussing, it would change the behavior of these command line arguments from preventing the "Add new torrent" dialog from opening to not preventing it. I think it's bad form to have a potential future version of the app change the behavior of command line arguments in that dramatic of a fashion.

So how about this: I'll add the TODO and change the `--skip-dialog` to be a tri-state bool. This already makes sense because there is actually an application-wide setting for whether the add torrent dialog should be opened, so being able to say `--skip-dialog=false` might be useful in some cases. For now I'll make a note in the command line usage text that using `--skip-dialog=false` in conjunction with the other command line arguments I've added is currently unsupported.

But that way if in a future version someone follows up on the TODO and changes `AddNewTorrentDialog` so that you can override its settings with these command line arguments, we can start allowing using them in conjunction with `--skip-dialog=false` to allow someone to pop open the dialog, override its settings, all while not change the existing behavior of the other command line arguments. Sound good?

evsh replied:
Sorry, I don't understand your point. I thought that it would be sufficient just add the option and in case of `--skip-dialog=false` and any of the new options set just abort and show that this combination of options is not yet implemented.

glassez replied:
Use TriStateBool for skipDialog is a really good idea. And, of course, we should implement overriding of AddNewTorrentDialog initial values (IMO, it should be done before these changes go to release).

briankendall replied:
Well being that I have a hard time implementing features that aren't quite finished or don't quite do *The Right Thing* I went ahead and modified `AddNewTorrentDialog` so that it works with my new command line options.

evsh commented on src/app/options.cpp:
```diff
@@ -235,6 +240,69 @@ namespace
+                return TriStateBool::False;
+            }
+            else {
+                //: Translation must use English "true" and "false"
```
Will this go into Transifex? Why don't you use the second argument of `QObject::tr()`?

briankendall replied:
The second argument of `QObject::tr()` is meant for disambiguation, like when you have two strings that are the same in English but might need to be translated differently in another language. You're supposed to use the `//:` comment notation for notes to the translator. See [this section of the Qt docs on translations](http://doc.qt.io/qt-5/i18n-source-translation.html#translator-comments)

evsh replied:
Does anybody know will this comment go into Transifex? @briankendall, maybe it is worth to write is simply as:
```c++
QObject::tr("Parameter '%1' must follow syntax '%1=%2'").arg(fullParameter()).arg(QLatin1String("<true|false>"));
```
?

briankendall replied:
Good idea

evsh commented on src/app/options.cpp:
```diff
@@ -235,6 +240,69 @@ namespace
+                return TriStateBool::False;
+            }
+            else {
+                //: Translation must use English "true" and "false"
```
Same here.

glassez commented:
I have one more (apparently, late) idea.
To not pollute command line with many parameters we can have some serialized form of AddTorrentParams (e.g. JSON) and pass only one parameter (e.g. `--add-torrent-params="<serialized_params>"`).

briankendall commented:
@glassez No, I think that's a bad idea. It makes specifying options for adding torrents much more difficult and obtuse, and completely breaks away from the conventions of the command line. It would frustrate and confuse users for seemingly no benefit.

Besides, there's nothing wrong with having lots of command line options. The whole point of them being there in the first place is that it gives you a convenient way to specify behavior from the command line, and you can use as many or as few as you want. With this PR, qBittorrent doesn't even have a lot of options in comparison to lots of other command line tools that are used constantly.

Chocobo1 commented:
>Besides, there's nothing wrong with having lots of command line options. 

And I started to wonder, is qBittorrent a cmdline tool, GUI, webUI or a Chimera?

briankendall commented:
Just updated again. Changes:

- Updated `AddNewTorrentDialog` so that it works with these command line options
- Changed the behavior of all the options other than `--skip-dialog` so that they have no effect on whether the new torrent dialog opens.
- `--skip-dialog` is now a tri-state bool option

LordNyriox commented on src/gui/addnewtorrentdialog.ui:
```diff
@@ -192,16 +192,6 @@
-      <item row="0" column="0">
-       <widget class="QCheckBox" name="startTorrentCheckBox_2">
-        <property name="text">
-         <string>Start torrent</string>
```
@briankendall:  Why was this section of the UI removed?

briankendall replied:
It was for a redundant "Start torrent" checkbox. It seemed to be the result of a copy/paste mistake someone made. It's not referenced anywhere in the code, and it obscured the real "Start torrent" checkbox, causing bugs.

LordNyriox replied:
@briankendall:  Ahhh.  OK.

MatrixView commented:
Hi guys, thanx for all the hard work! And even though I can't follow all your input, it seems you're really thinking this thing through.

Now, I don't want to be a party pooper, but..., can any of you tell me if these commandline options will also work if I don't specify a .torrent file, but a **.MAGNET file** (or a magnet-link)...?

As a scripting/automation person I would mostly like to use it in something similar like this:
**qBitTorrent.exe "D:\file.magnet" --category="Documentary" --save-path="D:\Docu" --add-paused=false --skip-dialog**

...where the D:\file.magnet (a text file) contains the magnet link(s).

briankendall commented:
Updated again. Fixed merge conflicts with master, and fixed the issue where specifying --first-and-last with a magnet link wouldn't work.

briankendall commented:
@MatrixView I don't think qBittorrent supports opening .magnet files like that. However, you can pass in the magnet link itself as part of the command line argument with those options, e.g.:

    qBitTorrent.exe "magnet:..." --category="Documentary" --save-path="D:\Docu" --add-paused=false --skip-dialog

MatrixView commented:
Ok, thx Brian!  qBitTorrent needs .magnet files (containing the magnet link) if you've setup watchfolder(s). But a direct magnet-link will be fine in most cases too...   except for magnet-links that contain special-chars like quotes and redirects chars such as: >, <, |, &
This would break the complete input to qBT due to the OS interpreting it differently. So in that case a .magnet file would be better.

briankendall commented:
@evsh I believe I've addressed everything from your review but it's still blocking merging. Can you clear that up?

glassez commented on src/app/application.cpp:
```diff
@@ -405,15 +405,62 @@ void Application::processParams(const QStringList &params)
+        // is false, meaning that the application setting should be overridden.
+        bool showDialogForThisTorrent = ((AddNewTorrentDialog::isEnabled() ||
+                                          skipTorrentDialog == TriStateBool::False) && 
+                                         skipTorrentDialog != TriStateBool::True);
```
It's better to split lines according to subexpressions:
```c++
bool showDialogForThisTorrent =
    ((AddNewTorrentDialog::isEnabled() || skipTorrentDialog == TriStateBool::False)
    && skipTorrentDialog != TriStateBool::True);
```

glassez commented on src/app/main.cpp:
```diff
@@ -72,6 +72,7 @@ Q_IMPORT_PLUGIN(QICOPlugin)
 #include "base/profile.h"
 #include "base/utils/misc.h"
 #include "base/preferences.h"
+#include "base/tristatebool.h"
```
Why it's here? It's unused at this file.

briankendall replied:
Accidentally left it in from a previous (now squashed) commit

glassez commented on src/app/options.cpp:
```diff
@@ -45,6 +45,10 @@
 
 #include "base/utils/misc.h"
 
+static const int USAGE_INDENTATION = 4;
```
`const` is static by default so you can omit `static` keyword in these lines.

evsh replied:
There is an unnamed namespace right below. These constants would fit there perfectly.

evsh commented on src/app/application.cpp:
```diff
@@ -405,15 +405,62 @@ void Application::processParams(const QStringList &params)
+        // One is when the application settings specify that it should be shown
+        // and skipTorrentDialog is undefined. The other is when skipTorrentDialog
+        // is false, meaning that the application setting should be overridden.
+        bool showDialogForThisTorrent = ((AddNewTorrentDialog::isEnabled() ||
```
`const bool`, please.

evsh commented on src/gui/addnewtorrentdialog.cpp:
```diff
@@ -205,6 +206,12 @@ void AddNewTorrentDialog::show(QString source, QWidget *parent)
 
+void AddNewTorrentDialog::show(QString source, QWidget *parent)
+{
+    BitTorrent::AddTorrentParams defaultParams;
```
Why do you need the named variable?

briankendall replied:
I like being explicit in the intent of a line of code, and using the named variable makes it more clear what I'm doing in this version of `AddNewTorrentDialog::show`. However I can see how it's being more explicit than is strictly necessary so I will remove the variable.

evsh commented on src/gui/addnewtorrentdialog.cpp:
```diff
@@ -416,6 +423,21 @@ void AddNewTorrentDialog::categoryChanged(int index)
+    const int existingIndex = indexOfSavePath(newPath);
+    if (existingIndex >= 0) {
+        ui->savePathComboBox->setCurrentIndex(existingIndex);
+        onSavePathChanged(existingIndex);
```
This can be moved after the `if` and you can remove line 437 then.

evsh replied:
Ah, you just borrowed this piece from the code below...

evsh commented on src/gui/addnewtorrentdialog.h:
```diff
@@ -112,6 +116,7 @@ private slots:
     QByteArray m_headerState;
     int m_oldIndex;
     QScopedPointer<TorrentFileGuard> m_torrentGuard;
+    BitTorrent::AddTorrentParams torrentParams;
```
If there is a member, it has to be initialised in the constructor, but not in `show()` function.

evsh commented on src/gui/addnewtorrentdialog.cpp:
```diff
@@ -621,32 +636,63 @@ void AddNewTorrentDialog::displayContentTreeMenu(const QPoint &)
+            ui->categoryComboBox->setCurrentIndex(index);
+        }
+        else {
+            ui->categoryComboBox->addItem(torrentParams.category);
```
And what if the category is absent? Say, user mistype it in the command line? I would prefer a program that tells me about that.

briankendall replied:
I thought the reasonable behavior in this case is for it to create a new category. Note that the user can type in any category they want into `categoryComboBox` and if it doesn't exist, qBittorrent will create it. So the command line argument should behave the same way as providing a value for the combo box. This means it's the responsibility of the user to get the category name right, but that's always been the case.

evsh replied:
I don't think this relation is correct now. In the GUI user actions are explicit, they know that the new category is created, but here this is done implicitly. To make these two cases equivalent, we should add to the command line help and note that category will be created if it does not exist.

evsh commented on src/gui/addnewtorrentdialog.h:
```diff
@@ -65,6 +66,7 @@ class AddNewTorrentDialog: public QDialog
     static bool isTopLevel();
     static void setTopLevel(bool value);
 
+    static void show(QString source, BitTorrent::AddTorrentParams &inParams, QWidget *parent);
```
Please no. Unparented dialogue windows are showed on the first screen always in Plasma and maybe in any X session regardless of the main window position.

glassez replied:
I suggested this to preserve signature similarity.
Then remove it from next `show()` too.

glassez commented on src/app/application.cpp:
```diff
@@ -405,15 +405,62 @@ void Application::processParams(const QStringList &params)
+        
+        if (param.startsWith("@savePath=")) {
+            torrentParams.savePath = param.mid(10);
+            continue;
```
It's not only my point of view.
The normal continuation/completion of the loop should be described in the appropriate place (as loop condition). `break` and `continue` instructions are for non-standard/emergency interruption of the loop (or loop iteration). Or, in exceptional cases, to avoid too much code (when other methods do not help or are ineffective).
I don't think that in the case of the use of `else if` this code will become less readable. Besides, it will become three times shorter!

evsh replied:
Nonsense, `break` and `continue` are different in this sense: `continue` has nothing to do with loop condition.

glassez replied:
Of course. I just said that `break` breaks loop, and `continue` breaks current loop iteration.

I don't insist on changing this.

evsh replied:
You are saying this as if everybody likes to read a loop body to the end to check whether there is anything after a sheet of `if`s.

glassez replied:
Since it consists only of these same blocks, I do not insist on the change. In another case it could be otherwise...

briankendall replied:
I'll keep as is. I agree with @evsh in this case that it is more readable with `continue`.

evsh replied:
> Since it consists only of these same blocks, I do not insist on the change. In another case it could be otherwise...

I'm pretty sure in that case the current fragment would evolve into a function containing only this switch-like structure.

glassez replied:
>I'm pretty sure in that case the current fragment would evolve into a function containing only this switch-like structure.

+1

glassez replied:
Maybe later.

briankendall commented on src/gui/addnewtorrentdialog.cpp:
```diff
@@ -621,32 +636,63 @@ void AddNewTorrentDialog::displayContentTreeMenu(const QPoint &)
+    }
+    else if (torrentParams.addPaused == TriStateBool::False) {
+        ui->startTorrentCheckBox->setChecked(true);
+    }
```
If `Undefined` then we leave the checkbox as is, since we fall back on the application setting. So nothing to do in that case, hence no `else` clause.

glassez replied:
You probably want to tell me: dude, look at another place where I set some value for this. I answer: it must be concentrated in one place.

glassez replied:
You need to place all this logic here and modify constructor to accept AddTorrentParams as parameter, then call this function from it.

briankendall commented on src/base/bittorrent/torrenthandle.cpp:
```diff
@@ -1222,8 +1226,12 @@ void TorrentHandle::toggleSequentialDownload()
 {
-    if (!hasMetadata()) return;
+    if (!hasMetadata()) {
+        m_needSetFirstLastPiecePriority = b;
```
@glassez I was trying to avoid using that flag as the canonical indicator as to whether a torrent has first/last piece priority set, since then we need to worry about it going out of sync with the actual state of the torrent. (Note that the current implementation of `hasFirstLastPiecePriority()` doesn't use such a flag but instead checks if the first and last pieces actually have the correct priority set, so there's nothing to get out of sync.)

However, if we need to store it in the resume data I do believe we'll need to use a flag that way.

Also, we need to be careful about setting the flag unconditionally, because we don't want the torrent to have first/last priority if it doesn't contain any media files that support it. So I'll maintain that behavior in how `m_hasFirstLastPiecePriority` is set.

briankendall replied:
Sequential download also needs to be saved with the resume data, so I'll fix that too.

briankendall commented on src/gui/addnewtorrentdialog.h:
```diff
@@ -38,6 +38,7 @@
 
 #include "base/bittorrent/infohash.h"
 #include "base/bittorrent/torrentinfo.h"
+#include "base/bittorrent/session.h"
```
Not possible due to `AddNewTorrentDialog` having a member variable of type `BitTorrent::AddTorrentParams`.

evsh replied:
We should probably refactor  `AddTorrentParams` into its own file...

glassez replied:
Or we can just get rid of this member. In fact we don't need it at all. We just use a parameter of this type to initialize the dialog, and then just create the object of this type for passing to the Session. Why we need to store it?

briankendall replied:
@glassez Well, we do need it, or something like it. Not every option in `BitTorrent::AddTorrentParams` is represented by UI in `AddNewTorrentDialog`. In particular, `firstLastPiecePriority` and `sequential`, both of which can be specified from the command line, do not have options in the dialog. But we still want them to be enabled if the user adds the torrent from the command line and specifies them using `--first-and-last` or `--sequential`, regardless of whether the dialog is invoked or not.

I opted to have `BitTorrent::AddTorrentParams` be a member variable because that was the safest way to ensure every option it specifies gets used when the torrent is actually added when `BitTorrent::Session::instance()->addTorrent()` is called by `accept()`. This is particularly relevant in case someone adds more fields to `BitTorrent::AddTorrentParams`, because those will be respected too -- it gives us a degree of future proofing and is less error prone.

That said, it is true that currently the only options we care about passing through the dialog *right now* that don't have a control in its UI are `firstLastPiecePriority` and `sequential`. In order to eliminate `m_torrentParams` we could add boolean member variables to `AddNewTorrentDialog` that just store those values, and then construct a new `BitTorrent::AddTorrentParams` in the `accept()` method.

However if the only reason to eliminate `m_torrentParams` is to remove an `#include` from the header file, I don't think it's a net win given that keeping it around has the benefits I listed above. But I'm not wedded to either approach. If you want me to eliminate `m_torrentParams` I'll do it.

glassez replied:
Then we really should place AddTorrentParams in its own header.

evsh replied:
Now it does not make any sense since AddTorrentParams class is in `base/bittorrent/addtorrentparams.h`

briankendall commented:
@evsh @glassez Just updated again to address the latest review. How're we doing now?

briankendall commented:
@evsh It is done.

evsh commented on src/app/options.cpp:
```diff
@@ -343,43 +489,84 @@ const QString& CommandLineParameterError::messageForUser() const
+    stream << CATEGORY_OPTION.usage(QObject::tr("name"))
+           << wrapText(QObject::tr("Assign torrents to category. If the category doesn't exist, it will be "
+                                   "created.")) << '\n';
+    stream << SEQUENTIAL_OPTION.usage() << wrapText(QObject::tr("Download torrents in sequential order")) << '\n';
```
torrents → files

briankendall commented:
@evsh Whoops. Good catch, now fixed.

briankendall commented:
Updated again because I forgot to uncomment some code in `AddNewTorrentDialog`'s initializer that properly set up the "create subfolder" checkbox.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -3372,8 +3372,14 @@ void Session::createTorrentHandle(const libt::torrent_handle &nativeHandle)
             torrent->resume(data.addForced);
+        
+        if (data.sequential)
+            torrent->setSequentialDownload(true);
```
It has done in TorrentHandle constructor.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -3372,8 +3372,14 @@ void Session::createTorrentHandle(const libt::torrent_handle &nativeHandle)
+            torrent->setSequentialDownload(true);
+        
+        if (data.firstLastPiecePriority)
+            torrent->setFirstLastPiecePriority(true);
```
It has done in TorrentHandle constructor.

glassez commented on src/base/bittorrent/torrenthandle.cpp:
```diff
@@ -1630,6 +1619,10 @@ void TorrentHandle::handleMetadataReceivedAlert(libt::metadata_received_alert *p
+    
+    // If first/last piece priority was specified when adding this torrent, set it up now that we can:
+    if (m_hasFirstLastPiecePriority)
+        setFirstLastPiecePriority(true);
```
How about Se

glassez replied:
How about Sequential here?

briankendall replied:
Not necessary. We can set sequential downloading on a torrent before we have its metadata. Not so for first / last piece priority.

glassez replied:
Then don't store it in resume data.

evsh replied:
But how to resume this mode then. Frankly speaking, I lost both of you in this discussion couple of iterations ago. IMO, the implementation has to be simple:
- add a member flag to `TorrentHandle`: `bool m_needApplyFirstLastPriorityMode;`
- in `TorrentHandle::setFirstLastPiecePriority()` if there are no metadata, set this flag, otherwise proceed as usual and reset the flag;
- in `TorrentHandle::handleMetadataReceivedAlert()` check the flag:  call  `setFirstLastPiecePriority(true);` when it is set.
- save the flag to resume data (probably only if it is set).

glassez replied:
I agree. 
This was done almost the same until the author decided to get into the inside.

>Update: the flag has to be serialised always, of course, otherwise we would read it as set from the resume file always.

Disagree. We should write this flag to resume data only if torrent hasn't metadata (as we do for addPaused/addForced).

briankendall replied:
@glassez I discovered that, similar to first/last piece priority, if a magnet link is added with sequential download but qBittorrent quits before its metadata is retrieved, then `sequential` is not saved. So I added some code to save `sequential` in the resume data when a torrent's metadata isn't retrieved yet to fix the bug.

glassez replied:
Ok. Then we need appropriate comment. Otherwise someone may break it.

glassez commented on src/base/bittorrent/torrenthandle.cpp:
```diff
@@ -729,30 +733,7 @@ bool TorrentHandle::isSequentialDownload() const
-    int last = m_nativeHandle.piece_priority(extremities.last());
-
-    return ((first == 7) && (last == 7));
+    return m_hasFirstLastPiecePriority;
```
It becomes incorrect when the user changes the priorities in a different way. 

Why did you change it, you're not going to do it? So you're never going to finish this PR.

briankendall replied:
What other way? The only way I'm aware of for these specific priorities to be set is to use the `setFirstLastPiecePriority` function.

Also, this was discussed earlier. I had originally introduced a member variable called `m_needsToSetFirstLastPiecePriority` to handle when a torrent is added with the `firstLastPiecePriority` option before it has downloaded its metadata, but I was asked to change it to `m_hasFirstLastPiecePriority`. That means that new flag is the canonical indicator as to whether a torrent has this option set, and it's no longer valid to check the actual priority of the pieces because it could be set before the torrent's metadata is available. What would you suggest I do otherwise?

evsh replied:
> What other way? 

Any manipulation with file priorities resets first-last priority mode, AFAIK. that is why the mode is tracked via piece priorities but not via a flag.

briankendall replied:
Ah I see, the resume data sets it as well. I don't think it makes sense for us to have this flag. I think my earlier approach of using `m_needsToSetFirstLastPiecePriority` was better.

evsh replied:
Oh, no, pardon me: my information is outdated.

glassez replied:
>Oh, no, pardon me: my information is outdated.

What do you mean?
>Any manipulation with file priorities resets first-last priority mode

This isn't true?

evsh replied:
You decoupled firs-last mode from sequential mode in  67e90d8d, but setting file priorities still reset piece priorities, right?

glassez replied:
>but setting file priorities still reset piece priorities, right?

AFAIK, yes!

glassez replied:
>but I was asked to change it to m_hasFirstLastPiecePriority. That means that new flag is the canonical indicator as to whether a torrent has this option set, and it's no longer valid to check the actual priority of the pieces because it could be set before the torrent's metadata is available.

You was asked to change field name only.
>I think my earlier approach of using m_needsToSetFirstLastPiecePriority was better.

Let it be. Otherwise this changes will never be released.
@evsh? IMO, we should omit changing internals here.

evsh replied:
See my comment below: https://github.com/qbittorrent/qBittorrent/pull/5532#discussion_r113232980

briankendall replied:
> You was asked to change field name only.

I see that now. I'm very used to variable names being inextricably linked to their purpose. I do think that I need to reverse this change as it's introduced new bugs, and go back to the method @evsh outlined. Sorry about that -- I swear I'm not trying to make extra work for all of you! I very much want to see this PR completed.

evsh replied:
> I very much want to see this PR completed.

Me too.

briankendall commented on src/base/bittorrent/session.cpp:
```diff
@@ -3372,8 +3372,14 @@ void Session::createTorrentHandle(const libt::torrent_handle &nativeHandle)
 
     if (data.resumed) {
-        if (fromMagnetUri && !data.addPaused)
+        if (fromMagnetUri && data.addPaused)
```
This is strange, and I'm not sure how I made that edit. Sorry! I've corrected it.

briankendall commented:
Okay, just updated again. *Hopefully* we're in a good state. Changes:

- Changed `m_firstLastPiecePriority` back to `m_needsToSetFirstLastPiecePriority`
- Restored `hasFirstLastPiecePriority()` back to how it was before but with one change: if there is no metadata, the value of `m_needsToSetFirstLastPiecePriority` is returned
- Moved logic for calling `setSequentialDownload` and `setFirstLastPiecePriority` out of `Session::createTorrentHandle` and into the constructor of `TorrentHandle`, with a comment explaining why it's written the way it is
- Add comment explaining need to store `sequential` and `firstLastPiecePriority` in resume data when there is no metadata

As near as I can tell, torrent resuming is working correctly again. Sequential and first/last piece priority can be properly set from the command line, and is saved and restored for magnet links without metadata, magnet links with metadata, and regular torrents.

The one slightly complicated bit I just added is the logic for calling `setSequentialDownload` and `setFirstLastPiecePriority` in the constructor of `TorrentHandle`. I put a comment explaining why I wrote it that way that I tried to make as clear and as accurate as possible. But the way it's written now is equivalent logic to how it was before, and I believe it is correct.

evsh commented on src/base/bittorrent/torrenthandle.cpp:
```diff
@@ -1241,6 +1259,7 @@ void TorrentHandle::setFirstLastPiecePriority(bool b)
         const QString ext = Utils::Fs::fileExtension(path);
         if (Utils::Misc::isPreviewable(ext) && (fp[index] > 0)) {
             qDebug() << "File" << path << "is previewable, toggle downloading of first/last pieces first";
+            hasPreviewableFile = true;
```
Can not find where is it used.

briankendall replied:
An artifact of an older version of the function. Will remove.

evsh commented on src/gui/addnewtorrentdialog.cpp:
```diff
@@ -91,15 +92,36 @@ AddNewTorrentDialog::AddNewTorrentDialog(QWidget *parent)
     connect(ui->browseButton, SIGNAL(clicked()), SLOT(browseButton_clicked()));
     ui->defaultSavePathCheckBox->setVisible(false); // Default path is selected by default
-    ui->createSubfolderCheckBox->setChecked(session->isCreateTorrentSubfolder());
+    ui->skipCheckingCheckBox->setChecked(m_torrentParams.skipChecking);
```
Quite a strange replacement.

briankendall replied:
Not technically a replacement. The code for setting `ui->createSubfolderCheckBox` is higher up in the function now so that it is grouped with similarly structured code for setting `ui->startTorrentCheckBox`, and `ui->skipCheckingCheckBox` is an addition now that `BitTorrent::AddTorrentParams` is a parameter the constructor.

I do agree with you that this would be much easier to follow if commits weren't being squashed. Let me know if you don't want me to do that any longer.

evsh commented on src/gui/addnewtorrentdialog.h:
```diff
@@ -87,7 +89,7 @@ private slots:
 
 private:
-    explicit AddNewTorrentDialog(QWidget *parent = 0);
+    explicit AddNewTorrentDialog(const BitTorrent::AddTorrentParams &inParams, QWidget *parent = 0);
```
Default value for parent does not make sense when the corresponding default gets removed from `show()`.

briankendall commented:
Updated. Changes:

- Reorganized code in `AddNewTorrentDialog` slightly so that replacements are easier to follow
- Removed default value to `parent` argument in constructor of `AddNewTorrentDialog`
- Removed defunct `hasPreviewableFile` variable in `TorrentHandle::setFirstLastPiecePriority()`

briankendall commented:
@glassez Updated to remove `m_needsToSetFirstLastPiecePriority = false;` line in `TorrentHandle::setFirstLastPiecePriority()`.

evsh commented:
> @glassez Updated to remove m_needsToSetFirstLastPiecePriority = false; line in TorrentHandle::setFirstLastPiecePriority().

Why?

glassez commented:
@evsh, you apparently missed my comment about this. If we reach this line then torrent has its metadata and `m_needsToSetFirstLastPiecePriority` is already false.

evsh commented:
@glassez, can't understand how that can be true when we came from `handleMetadataReceivedAlert()`.

glassez commented:
Oops... My mistake.
I was confused by the fact that we have it at every function call, and must do it only once, when metadata is loaded. That is, it is more logical to do it in handleMetadataReceivedAlert.

briankendall commented:
@glassez @evsh Do you want me to make any other changes regarding `m_needsToSetFirstLastPiecePriority = false;`? I think it's okay if we keep that line removed because as soon as a `TorrentHandle` has metadata we never use the flag again.

evsh commented:
@briankendall, I want that change, in order to make `m_needsToSetFirstLastPiecePriority` value correspond to the variable name and to make code intention clean at that place.

briankendall commented:
@evsh Ok, done. I also rebased to fix another merge conflict.

glassez commented:
@evsh, IMO, there are no large changes in the core and we can not wait for @sledgehammer999 approval, isn't it?

evsh commented:
He was around these days. Let's wait until tomorrow?

glassez commented:
>He was around these days. Let's wait until tomorrow?

Ok.

evsh commented:
Apparently, he is not around anymore. Then we need an approval from @Chocobo1.

evsh commented:
@Chocobo1, could you review this PR, please?

Chocobo1 commented on src/app/options.cpp:
```diff
@@ -264,9 +338,60 @@ QBtCommandLineParameters::QBtCommandLineParameters(const QProcessEnvironment &en
+        result.append(QString("@savePath=%1").arg(savePath));
+
+    if (addPaused == TriStateBool::True) {
+        result.append(QString("@addPaused=1"));
```
no need of `QString()` here.

Chocobo1 commented on src/app/options.cpp:
```diff
@@ -264,9 +338,60 @@ QBtCommandLineParameters::QBtCommandLineParameters(const QProcessEnvironment &en
+        result.append(QString("@addPaused=1"));
+    }
+    else if (addPaused == TriStateBool::False) {
+        result.append(QString("@addPaused=0"));
```
no need of `QString()` here.

Chocobo1 commented on src/app/options.cpp:
```diff
@@ -235,6 +240,68 @@ namespace
+        }
+    };
+
+    inline bool operator==(const QString &s, const TriStateBoolOption &o)
```
if it's not critical, please remove `inline`, I prefer to trust the compiler decision.

Chocobo1 commented on src/app/options.cpp:
```diff
@@ -264,9 +338,60 @@ QBtCommandLineParameters::QBtCommandLineParameters(const QProcessEnvironment &en
+        result.append("@firstLastPiecePriority");
+
+    if (skipDialog == TriStateBool::True) {
+        result.append(QString("@skipDialog=1"));
```
no need of `QString()` here and below.

Chocobo1 commented on src/base/bittorrent/torrenthandle.cpp:
```diff
@@ -729,7 +741,9 @@ bool TorrentHandle::isSequentialDownload() const
 bool TorrentHandle::hasFirstLastPiecePriority() const
 {
-    if (!hasMetadata()) return false;
+    if (!hasMetadata()) {
```
minor thing, you can remove the curly braces here if you want.

Chocobo1 commented on src/app/options.h:
```diff
@@ -57,9 +59,17 @@ struct QBtCommandLineParameters
+    QString category;
+    bool sequential;
+    bool firstLastPiecePriority;
+    TriStateBool skipDialog;
```
this is causing some padding bytes in the struct.
if you want to fix it (as this is not directly related to this PR), you can follow the style in `gui/advancedsettings.h`.

briankendall commented:
@Chocobo1 Rebased and made the changes you requested

evsh commented on src/app/application.cpp:
```diff
@@ -407,15 +407,62 @@ void Application::processParams(const QStringList &params)
+
+        // Process strings indicating options specified by the user.
+        
+        if (param.startsWith("@savePath=")) {
```
If you don't mind to spare another couple of minutes, please change this now to
```c++
if (param.startsWith(QLatin1String("@savePath="))) {
```

briankendall replied:
I'll make this change in a little while when I'm free, but I'm curious: what's the advantage of using QLatin1String instead of a string literal? Does it make any difference when you're just dealing with ASCII characters?

evsh replied:
> I'll make this change in a little while when I'm free…

Thanks you!
>…but I'm curious: what's the advantage of using QLatin1String instead of a string literal? Does it make any difference when you're just dealing with ASCII characters?

Apart from indicating for a reader an untranslatable string literal, it also gives a hint to QString that conversion to Unicode 16 has to be done from Latin-1 but not from UTF-8, which is more efficient.

evsh replied:
@briankendall: you forgot to add `QLatin1String` here.

briankendall replied:
@evsh Got it

briankendall commented:
@evsh Made the changes

sledgehammer999 commented on src/app/options.cpp:
```diff
@@ -343,43 +489,84 @@ const QString& CommandLineParameterError::messageForUser() const
     stream << QObject::tr("Usage:") << '\n';
 #ifndef Q_OS_WIN
-    stream << '\t' << prgName << " [options] [(<filename> | <url>)...]" << '\n';
+    stream << indentation << prgName << QLatin1String(" [options] [(<filename> | <url>)...]") << '\n';
```
Does anyone know why we don't print this on Windows? It seems like a mistake.

briankendall replied:
I didn't write the original version of that line, so I don't know. From my perspective I don't see any reason not to include it in Windows, though. If you'd like me to remove the `#ifndef Q_OS_WIN` I can do that.

sledgehammer999 replied:
If it isn't too much trouble for you, please do.

LordNyriox commented on src/base/bittorrent/torrenthandle.cpp:
```diff
@@ -1508,6 +1524,11 @@ void TorrentHandle::handleSaveResumeDataAlert(libtorrent::save_resume_data_alert
         resumeData["qBt-forced"] = isForced();
+        // Both firstLastPiecePriority and sequential need to be stored in the
+        // resume data if there is no metadata, otherwise they won't be
+        // restored if qBittorrent quites before the metadata is retrieved:
```
@sledgehammer999, @briankendall:  Actually, `is retrieved` is correct ("metadata" is treated as singular in most cases of English syntax).  

I agree with you about `quite`/`quit`, however.

sledgehammer999 replied:
`metadata` should be treated the same as `data`. They are in plural. And I like my oldschool (and correct) English!!!
1. http://www.dictionary.com/browse/data
2. http://www.dictionary.com/browse/metadata

briankendall replied:
I'm generally a descriptivist when it comes to language and I think at this point "data" and "metadata" can both be used as singular words as it's undeniably become a common usage. But regardless I'll make the change.

evsh replied:
A datum? Is it not used nova days?

briankendall commented on src/app/application.cpp:
```diff
@@ -407,15 +407,62 @@ void Application::processParams(const QStringList &params)
+        // is false, meaning that the application setting should be overridden.
+        const bool showDialogForThisTorrent =
+            ((AddNewTorrentDialog::isEnabled() || skipTorrentDialog == TriStateBool::False)
+             && skipTorrentDialog != TriStateBool::True);
```
I agree!

briankendall commented:
@sledgehammer999 Made the requested changes

sledgehammer999 commented:
Thx for PR.

briankendall commented:
Hooray! 🎉👍 Thanks everyone!

MatrixView commented:
Yeah!!! Thanx to all coders!  Can't wait for the release/version!

evsh commented:
Thank you, @briankendall, thanks to all the reviewers!
